### PR TITLE
Quarterdeck accept user invites

### PR DESCRIPTION
**Pull request by pdeziel**
### Scope of changes

This updates the `Login` and `Register` endpoints in Quarterdeck to accept user invite tokens.

For Login, only one of `org_id` and `invite_token` can be provided.

For Register, only one of `project_id` and `invite_token` can be provided.

Fixes SC-15468

### Type of change

- [x] new feature
- [ ] bug fix
- [ ] documentation
- [ ] testing
- [ ] technical debt
- [ ] other (describe)

### Acceptance criteria

Describe how reviewers can test this change to be sure that it works correctly. Add a checklist if possible.

### Author checklist

- [x] I have manually tested the change and/or added automation in the form of unit tests or integration tests
- [ ]  I have updated the dependencies list
- [ ]  I have recompiled and included new protocol buffers to reflect changes I made
- [ ]  I have added new test fixtures as needed to support added tests
- [x]  Check this box if a reviewer can merge this pull request after approval (leave it unchecked if you want to do it yourself)
- [x]  I have moved the associated Shortcut story to "Ready for Review"

### Reviewer(s) checklist

- [ ] Any new user-facing content that has been added for this PR has been QA'ed to ensure correct grammar, spelling, and understandability.
- [ ] Are there any TODOs in this PR that should be turned into stories?
- [ ] Does it make sense for invite verification to be integrated into Login and Register?

### CODE REVIEW

File: pkg/quarterdeck/api/v1/api.go [hunk 1]
```diff
@@ -81,6 +81,7 @@ type PageQuery struct {
 
 type RegisterRequest struct {
 	ProjectID    string `json:"project_id"`
+	InviteToken  string `json:"invite_token"`
```
[the remark this answers is not in the document]
Comment (bbengfort) [reply pkg/quarterdeck/api/v1/api.go:84]: Ok, let's make sure we discuss this with Cletus at the UI sync today.

File: pkg/quarterdeck/auth.go [hunk 2]
```diff
@@ -89,10 +89,53 @@ func (s *Server) Register(c *gin.Context) {
...
+				return
+			}
+
+			sentry.Error(c).Err(err).Msg("could not retrieve invite token")
+			c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not process registration"))
+			return
+		}
+
+		// Verify that the invite token is for this email address
+		// TODO: Should we allow users to register with a different email address?
```
Comment (bbengfort) [inline pkg/quarterdeck/auth.go:119]: Another good question we should discuss with the group.

**Review by bbengfort — APPROVED** [review]
Big PR! I like the idea of integrating the invite token with the existing links; we'll discuss with Cletus today about adding hidden fields in the form for the invite token. 

The logic for login and register is now significantly more complicated - what coverage do you have on the tests? We may want to create a follow up story to list all the possible test cases and security exceptions we're working with and ensure that we have a test case for each of them.

**Comment by pdeziel** [comment]
> Big PR! I like the idea of integrating the invite token with the existing links; we'll discuss with Cletus today about adding hidden fields in the form for the invite token.
> 
> The logic for login and register is now significantly more complicated - what coverage do you have on the tests? We may want to create a follow up story to list all the possible test cases and security exceptions we're working with and ensure that we have a test case for each of them.

Yeah, I'm pretty confident about the main path and most of the "not found" error cases, however I think it's a good idea to enumerate all the possible test cases to make sure the error handling is correct so I'll create a story for that.

**Pull request merged**